csrc/preseg_passes/propagate_shardings.cpp

```diff
@@ -19,6 +19,182 @@
+}
+
+void splitLike(TensorView* tv, int64_t axis, Split* ref_split) {
+  NVF_ERROR(
+      isSplitDivisible(tv->axis(axis), ref_split),
+      "Require the sharded ID to be divisible by the split factor. Got: ",
+      tv->axis(axis),
+      " and split factor: ",
+      ref_split->factor());
+  tv->outer_split(axis, ref_split->factor());
+}
+
+// Returns the number of DID axis on reshaped ids that were propagated to the
+// consumer.
+void shardViewOp(ViewOp* view_op, int64_t& did_pos) {
+  // This implementation asserts that only one sharding is applied on the
+  // reshaped ids. Inner split is not supported. The cases are:
+  // 1. Split reshape: [h] -> [a, h/a]. Sharding on h is applied to a in
+  // consumer.
+  // 2. Merge reshape: [a, h/a] -> [h]. Sharding on a is applied to h in
+  // consumer.
+  // 3. Multiple splits or merge reshapes: [x, y, z] -> [xyz]. Sharding on x and
+  // xyz. Similarly for the corresponding split reshape.
+  // 4. Independent splits or merge reshapes: [w, x, y, z] -> [wx, yz]. Sharding
+  // is on w and y. In the consumer, it is applied to wx and yz.
+  // An improvement is to support mult-levels of sharding (not a real case yet)
+  // if they are all outer splits. For example: For the reshape [h] -> [a, h/a]
+  // where the h is sharded twice: [h] -> [cp, h/cp] -> [cp, tp, h/(cp*tp)]
+
+  // A more general approach maybe to "undo" the reshape (reverse transforms
+  // from root to logical domain), followed by simplification of the consumer
+  // loop domain to move DID upwards.
+
+  TensorView* producer = view_op->in();
+  TensorView* consumer = view_op->out();
+
+  const std::unordered_map<IterDomain*, IterDomain*>& c2p =
+      PairwiseLogicalDomainMap(producer, consumer).mapConsumerToProducer();
+  const std::unordered_map<IterDomain*, IterDomain*>& p2c =
+      PairwiseLogicalDomainMap(producer, consumer).mapProducerToConsumer();
+  auto [p_logical_reshaped_ids, c_root_reshaped_ids] =
+      getReshapedIds(view_op, c2p);
+
+  auto p_loop_domain = producer->getLoopDomain();
+
+  // Track number of DID axis on reshaped ids that were propagated to the
+  // consumer. These will not be included in TransformPropagator.
+  int64_t num_reshape_shardings = 0;
+
+  for (auto idx : c10::irange(did_pos)) {
+    IterDomain* p_did = p_loop_domain.at(idx);
+    NVF_ERROR(p_did->isDeviceDim());
+
+    auto p_transforms = DependencyCheck::getAllExprsBetween(
+        {p_logical_reshaped_ids.begin(), p_logical_reshaped_ids.end()},
+        {p_loop_domain.at(idx)});
+
+    if (p_transforms.empty()) {
+      // This device axis is not on reshaped ids.
+      continue;
+    }
+
+    if (p_transforms.size() > 1) {
+      // This reshape has been transformed.
+      // This can happen, for example, when there is a consumer-to-producer
+      // propagation before this pass. We will attempt using TransformReplay
+      // to propagate the sharding.
+      // TODO: Support multi-level shardings for cp + tp.
+      continue;
```

Comment:
Do we run into this already? 

I ask this because the code below seems to handle **multiple** outer-splits pretty well with slight modifications.

Reply:
No, this is not a practical case yet.
Handling a series of outer splits should be an easier extension. I wanted to do it in a separate PR where I can have tests with 2D sharding to exercise this.
If it is preferable to have it one PR, I can add that.

Reply:
> If it is preferable to have it one PR, I can add that.

I'd give this a try. I think this leads to more general **and** simpler code (one less `if` statement).

Reply:
I attempted this for:
1. 2D sharding - I was blocked by Issue #4440
2. Multiple divisble outer-splits. For example: `[h]->[a,h/a]` where `h` is first outer split by some constant factor `k` and then `k` by `DIDx(d)`. I ran into an issue in the pointwise scheduler which I will look into separately.

I reorganized the implementation still to be easily extensible for these cases -- L162-L183 can be a for loop applying multiple producer transformers to the consumer iteratively.

Reply:
Thanks for trying! I realized I was unclear. I didn't mean to ask for more tests. I was asking, implementation wise, instead of 
```
if len > 1:
  error out
process one step
```
do
```
for _ in len:
  process one step
```
The latter is both more general and simpler. 

Anyhow, I'm fine with the logic as is.

Reply:
L137-139 should do something similar in the updated code. Can you check once if you're viewing the latest file?

```
    NVF_ERROR(
        p_transforms.size() == 1 && p_transforms.front()->isA<Split>(),
        "Expected a split transform producing the device id.");
```